keyserver/listener.go

```diff
@@ -0,0 +1,115 @@
+package keyserver
+
+import (
+	"encoding/json"
+	"io"
+	"log"
+	"net"
+	"time"
+
+	. "github.com/coniks-sys/coniks-go/protocol"
+)
+
+func (server *ConiksServer) listenForRequests(ln net.Listener, handler func(msg []byte) ([]byte, error)) {
+	defer ln.Close()
+	go func() {
+		<-server.stop
+		ln.Close()
+	}()
+
+	for {
+		conn, err := ln.Accept()
+		if err != nil {
+			select {
+			case <-server.stop:
+				return
+			default:
+				log.Printf("accept client: %s", err)
+				continue
+			}
+		}
+		server.waitStop.Add(1)
+		go func() {
+			server.acceptClient(conn, handler)
+			server.waitStop.Done()
+		}()
+	}
+}
+
+func (server *ConiksServer) acceptClient(conn net.Conn, handler func(msg []byte) ([]byte, error)) {
+	defer conn.Close()
+	closed := make(chan struct{})
+	defer close(closed)
+	go func() {
+		select {
+		case <-server.stop:
+			conn.Close()
+		case <-closed:
+		}
+	}()
+
+	conn.SetDeadline(time.Now().Add(5 * time.Second))
+
+	// handle request
+	buf := make([]byte, 4096)
+	for {
+		n, err := conn.Read(buf)
+		if err != nil {
+			select {
+			case <-server.stop:
+				return
+			default:
+				if err != io.EOF {
+					log.Printf("client read %v: %v", conn.RemoteAddr(), err)
+				}
+				return
+			}
+		}
+
+		res, err := handler(buf[:n])
+		if err != nil {
+			log.Printf("client handle %v: %v", conn.RemoteAddr(), err)
+		}
+
+		n, err = conn.Write([]byte(res))
+		if err != nil {
+			select {
+			case <-server.stop:
+				return
+			default:
+				log.Printf("client write %v: %v", conn.RemoteAddr(), err)
+				return
+			}
+		}
+	}
+}
+
+func malformedClientMsg() []byte {
+	res, err := MarshalErrorResponse(ErrorMalformedClientMessage)
+	if err != nil {
+		panic(err)
+	}
+	return res
+}
+
+func (server *ConiksServer) makeHandler(acceptableTypes map[int]bool) func(msg []byte) ([]byte, error) {
+	return func(msg []byte) ([]byte, error) {
+		// get request message
+		req, err := UnmarshalRequest(msg)
+		if err != nil {
+			return malformedClientMsg(), err
+		}
+
+		switch true {
+		case req.Type == RegistrationType && acceptableTypes[RegistrationType]:
+			response, e := server.dir.HandleOps(req)
```

Comment:
Yes, that's the body that I'd like to see.

> In case of KeyLookup and KeyLookupInEpoch we probably just need to use server.RLock() instead of server.Lock(). It's sad :(

I think adding switches here to apply locks might be ok.

```
switch req.Type {
  default:
    server.Lock()
}
response, e := server.dir.HandleOps(req)
switch req.Type {
  default:
    server.UnLock()
}
```

Either that, or you move the mutex to the Directory, since that has knowledge of what actually needs the locking (or, even the Merkletree).